### PR TITLE
Transitive reduction of directected acyclic graphs

### PR DESCRIPTION
**Pull request by symphorien**
Here is an implementation of the transitive reduction and closure of directed acyclic graphs.
The source of the algorithm is [On the calculation of transitive reduction-closure of orders](https://www.sciencedirect.com/science/article/pii/0012365X9390164O) by Habib, Morvan and Rampon.

### Scope

The fact that it is limited to directed acyclic graphs is "normal" because transitive reduction is not well defined on graphs with a cycle. You can still define a set of minimal transitive reductions on cyclic graphs, though, and to obtain one, the algorithm is as follows:
* condense the graph
* on each scc component, an algorithm exists in linear time to compute *a* reduction
* use the algorithm of the implemented in this PR on the condensation
* combine the results of the two above steps in a transitive reduction of the original graph.

Therefore, if someone wants to extend petgraph to have transitive reduction for all directed graphs, then this algo will likely be reused and this is not a waste of time.

### Why is the PR so big ?

The algorithm needs a very specific data structure: an adjacency list where the nodes and each successor lists are toposorted. Practically speaking, this means that you have to copy the graph to sort it, so the size of the copy should be a small as possible. (Same apply for the result) So I added a new graph datastructure, `adj::List` which is *just* an adjacency list (only successors, no predecessors, like `Graph`). I implemented all the traits I could for this data structure, and this is why the pull request is so big.

### Why are there no node weights on `adj::List` ?
I have been bitten several times in the past by the following "shortcoming" of petgraph:
I want to use `EdgeFiltered` on a graph to visit it, but also change the node weights while visiting it.
But creating an EdgeFiltered borrows the graph, so it is impossible. The underlying problem is that one cannot borrow the weights separately from the nodes.
For the case of `adj::List`, I thought we could add something like
```rust
impl<N, E, W, Ix:IndexType> DataMap for (W, List<E, Ix>)
where W: Index<Ix, Output = N> {
...
}
```
ie you carry the graph and node weights separately, but by the magic of traits, a pair of these can be used as a fully-fledged graph. 

---

The PR is in good shape, but some small things could get some improvements. I would like you to confirm your interest in this work before I invest more time in it, though.

### CODE REVIEW

**Comment by XVilka** [comment]
Please rebase on top of the master to check with Travis properly.

**Comment by symphorien** [comment]
Done.

File: src/csr.rs
Comment (waywardmonkeys) [inline src/csr.rs]: I made an error here in my merge.

**Comment by symphorien** [comment]
I don't really see the point of the last commit, @waywardmonkeys ?

File: src/algo/tred.rs
Comment (jmcomets) [inline src/algo/tred.rs]: I believe a type alias for `List<(), Ix>` could make things a bit clearer. Maybe you could have other versions of these functions taking a node/edge mapping function, with the current functions passing `|_| ()` for both.

**Comment by waywardmonkeys** [comment]
@symphorien I was rebasing it forward to fix the merge conflicts ... but I made one typo (a missing `)`).

File: tests/list.rs
Comment (symphorien) [reply tests/list.rs]: Well spotted!

File: src/visit/mod.rs
Comment (symphorien) [reply src/visit/mod.rs]: They became unnecessary (and conflicting) because of https://github.com/petgraph/petgraph/commit/30fb481d7971732b1a025acb1c5b8ecc697b0fef#diff-ad1a163e094362a54511320c14281ffc

File: src/algo/tred.rs
Comment (symphorien) [reply src/algo/tred.rs]: done

File: src/algo/tred.rs
Comment (symphorien) [reply src/algo/tred.rs]: I added `UnweighedList` as an alias for `List<()>`.

**Comment by symphorien** [comment]
Most review comments should be addressed by now.

**Comment by jmcomets** [comment]
Looks great! I need some time to review the implementation of `List` and the new algorithms before merging this. I'll get back to you if there's anything else. 😉

**Comment by symphorien** [comment]
Any update ?

**Comment by symphorien** [comment]
Now that 0.5 is released, is there any interest in merging this PR ?
If so I will try to solve the merge conflict.

**Comment by symphorien** [comment]
bump

**Comment by symphorien** [comment]
I solved the merge conflicts. Can someone review this PR ? 
cc @XVilka @waywardmonkeys @jmcomets

**Review by XVilka — APPROVED** [review]
Thank you, looks good. Approving, if nobody objects in a few days, lets merge it.

**Pull request merged**
